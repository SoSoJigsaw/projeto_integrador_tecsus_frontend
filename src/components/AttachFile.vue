<template>
  <transition name="fade" mode="out-in">
    <div v-if="isVisible" class="main">
      <form @submit.prevent="sendData" ref="form">
        <div class="dropzone-container" @dragover="dragover" @dragleave="dragleave" @drop="drop">
          <input
            type="file"
            multiple
            name="file"
            id="fileInput"
            class="hidden-input"
            @change="onChange"
            ref="file"
            accept=".csv"
          />

          <label for="fileInput" class="file-label">
            <div v-if="isDragging" class="file-label-text">Drag files here</div>
            <div v-else class="file-label-text">
              Arraste os CSVs aqui ou <u>clique aqui</u> para fazer o download
            </div>
          </label>

          <div class="preview-container mt-4" v-if="files.length">
            <div v-for="file in files" :key="file.name" class="preview-card">
              <div>
                <img class="preview-img" :src="excel" />
                <p>
                  {{ file.name }}
                  ({{ Math.round(file.size / 1000) + 'kb' }})
                </p>
              </div>
              <div>
                <button
                  class="ml-2"
                  type="button"
                  @click="remove(files.indexOf(file))"
                  title="Remove file"
                >
                  <b>×</b>
                </button>
              </div>
            </div>
          </div>
        </div>
        <button v-if="this.files.length !== 0" type="submit" class="submit">Enviar</button>
      </form>
    </div>
  </transition>
</template>

<script lang="ts">
import axios from 'axios'

export default {
  data() {
    return {
      isDragging: false,
      files: [],
      filesJSON: [],
      excel: 'src/assets/icons/csv.png',
      isVisible: false
    }
  },

  mounted() {
    if (this.$refs.file) {
      this.$refs.file.addEventListener('change', (event) => {
        this.files = Array.from(event.target.files)
      })
    }

    setTimeout(() => {
      this.isVisible = true
    }, 1000)
  },

  methods: {
    onChange() {
      this.files = this.$refs.file.files

      for (let i = 0; i < this.files.length; i++) {
        const file = this.files[i]
        const reader = new FileReader()
        reader.onload = (e) => {
          const data = e.target.result
          const workbook = XLSX.read(data, { type: 'binary' })
          const worksheet = workbook.Sheets[workbook.SheetNames[0]]
          const json = XLSX.utils.sheet_to_json(worksheet)

          // objeto de mapeamento de nomes de chave original para o novo nome de chave
          const keyMap = {
            'Boletim de serviço': 'bulletin_service',
            Status: 'Status'
          }

          const header = Object.keys(json[0]) // obter as chaves do primeiro objeto como o cabeçalho
          const chassisValue = worksheet['C2'].v // obter o valor da célula C2 (segunda linha, terceira coluna)
          const newData = []
          json.forEach((row, index) => {
            // copiar o valor de "Chassis" para todas as linhas
            const newRow = {}
            header.forEach((key) => {
              newRow[keyMap[key] || key] = row[key]
            })
            if (index === 0) {
              // remover a chave "Chassis " com espaço
              delete newRow['Chassis ']
              delete newRow['Chassis']
            }
            newRow['chassis'] = chassisValue
            newData.push(newRow)
          })
          this.filesJSON.push(newData)
        }
        reader.readAsBinaryString(file)
      }
    },

    dragover(e) {
      e.preventDefault()
      this.isDragging = true
    },

    dragleave() {
      this.isDragging = false
    },

    drop(e) {
      e.preventDefault()
      this.$refs.file.files = e.dataTransfer.files
      this.onChange()
      this.isDragging = false
    },

    remove(i) {
      this.files.splice(i, 1)
    },

    sendData() {
      if (this.files.length > 0) {
        this.filesJSON.forEach((data) => {
          const jsonData = JSON.stringify(data)

          console.log(jsonData)

          // enviar arquivos para o servidor
          axios.post('http://localhost:8080/register/bulletin', jsonData, {
            headers: {
              'Content-Type': 'application/json'
            }
          })
        })

        // resetar o valor do input
        this.$refs.file.value = null
        this.files = []
        this.filesJSON = []

        alert('Data sent to database succesfully!')
      }
    }
  }
}
</script>

<style>
@import '../assets/css/base.css';

.main {
  display: flex;
  flex-grow: 1;
  align-items: center;
  justify-content: center;
  text-align: center;
}

.dropzone-container {
  padding: 7rem;
  background: #f7fafc;
  border: 1px solid #e2e8f0;
  border-radius: 10px;
  box-shadow: 2px 2px 20px 10px var(--silver);
}

.hidden-input {
  opacity: 0;
  overflow: hidden;
  position: absolute;
  width: 1px;
  height: 1px;
}

.file-label {
  font-size: 20px;
  display: block;
  cursor: pointer;
}

.file-label-text {
  color: black;
  font-weight: var(--medium);
  font-size: 25px;
}

.preview-container {
  display: grid;
  grid-template-columns: 50% 50%;
  grid-gap: 10px;
  margin-top: 2rem;
}

.preview-card {
  display: flex;
  border: 1px solid var(--platinum);
  border-radius: 10px;
  background-color: white;
  padding: 5px;
  margin-left: 5px;
  box-shadow: 2px 2px var(--silver);
}

.preview-card p {
  font-weight: var(--light);
  color: var(--azul-principal);
  text-align: center;
}

.preview-img {
  width: 50px;
  height: 50px;
  border-radius: 5px;
  border: 1px solid white;
  background-color: white;
}

.submit {
  background-color: aquamarine;
  color: #ffff;
  font-size: 14pt;
  border-radius: 5px;
  border: none;
  cursor: pointer;
  width: 100px;
  height: 40px;
  transition: all 1s;
  margin-top: 25px;
  box-shadow: 2px 2px 20px 10px var(--silver);
}

button:hover {
  opacity: 0.8;
}

.ml-2 {
  background-color: transparent;
  border: none;
}

.ml-2 b {
  font-size: 20px;
  color: black;
}

/* --------------- Media Queries -------------------- */

/* Estilos para tablet */
@media only screen and (min-width: 768px) and (max-width: 1023px) {
  .main {
    margin-left: 25px;
    margin-right: 25px;
    margin-top: 50px;
    margin-bottom: 50px;
    padding-bottom: 80px;
  }

  .dropzone-container {
    padding: 4rem;
    box-shadow: 2px 2px 20px 10px var(--silver);
    margin-right: 40px;
    margin-left: 40px;
  }

  .hidden-input {
    width: 1px;
    height: 1px;
  }

  .file-label {
    font-size: 20px;
  }

  .file-label-text {
    font-size: 25px;
  }

  .preview-container {
    display: grid;
    grid-template-columns: 50% 50%;
    grid-gap: 10px;
    margin-top: 2rem;
  }

  .preview-card {
    padding: 5px;
    margin-left: 5px;
    box-shadow: 2px 2px var(--silver);
  }

  .preview-card p {
  }

  .preview-img {
    width: 50px;
    height: 50px;
  }

  .submit {
    font-size: 14pt;
    width: 100px;
    height: 40px;
    margin-top: 25px;
    box-shadow: 2px 2px 20px 10px var(--silver);
  }

  .ml-2 b {
    font-size: 20px;
  }
}

/* Estilos para mobile */
@media only screen and (max-width: 767px) {
  .main {
    margin-left: 10px;
    margin-right: 10px;
    margin-top: 40px;
    margin-bottom: 50px;
    width: 95%;
    padding-bottom: 80px;
  }

  .dropzone-container {
    padding: 1rem;
    box-shadow: 2px 2px 50px 5px var(--silver);
    margin-right: 10px;
    margin-left: 10px;
  }

  .hidden-input {
    width: 1px;
    height: 1px;
  }

  .file-label {
    font-size: 20px;
  }

  .file-label-text {
    font-size: 18px;
  }

  .preview-container {
    display: grid;
    grid-template-columns: 47% 47%;
    grid-gap: 10px;
    margin-top: 2rem;
  }

  .preview-card {
    padding: 5px;
    margin-left: 5px;
    box-shadow: 2px 2px var(--silver);
  }

  .preview-card p {
  }

  .preview-img {
    width: 40px;
    height: 40px;
  }

  .submit {
    font-size: 12pt;
    width: 90px;
    height: 40px;
    margin-top: 25px;
    box-shadow: 2px 2px 20px 10px var(--silver);
  }

  .ml-2 b {
    font-size: 20px;
  }
}
</style>
