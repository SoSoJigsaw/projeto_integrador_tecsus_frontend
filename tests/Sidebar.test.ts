import { describe, it, expect, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import AttachFile from '@/components/Sidebar.vue'

it('should mount the component correctly', () => {
  const wrapper = mount(Sidebar)
  expect(wrapper.exists()).toBe(true)
})

it('should start with isVisible set to false', () => {
  const wrapper = mount(Sidebar)
  expect(wrapper.vm.isVisible).toBe(false)
})

it('should display the sidebar after 1 second', async () => {
  vi.useFakeTimers()
  const wrapper = mount(Sidebar)
  vi.advanceTimersByTime(1000)
  await wrapper.vm.$nextTick()
  expect(wrapper.find('.sidebar').isVisible()).toBe(true)
  vi.useRealTimers()
})

it('should redirect to specified route when link is clicked', async () => {
  const pushSpy = vi.fn()
  const wrapper = mount(Sidebar, {
    global: {
      mocks: {
        $router: {
          push: pushSpy
        }
      }
    }
  })
  await wrapper.findAllComponents(RouterLink).at(0).trigger('click.prevent')
  expect(pushSpy).toHaveBeenCalled() // Adicione verificações específicas para cada rota se necessário
})

it('should change animation delay after animation iteration', async () => {
  const wrapper = mount(Sidebar)
  await wrapper.find('.logo_img').trigger('animationiteration')
  expect(wrapper.vm.animationDelay).toBe(true)
})

it('should display all navigation icons', () => {
  const wrapper = mount(Sidebar)
  const icons = wrapper.findAll('i')
  expect(icons.length).toBeGreaterThan(3) // Verifica se todos os ícones esperados estão presentes
})

it('should change style on hover over links', async () => {
  const wrapper = mount(Sidebar)
  const link = wrapper.find('.links a')
  await link.trigger('mouseover')
  expect(link.element.style.opacity).toBeLessThan(1) // Assume uma mudança de opacidade ao passar o mouse
})

it('should apply active style to the currently active route', async () => {
  const wrapper = mount(Sidebar, {
    global: {
      stubs: {
        RouterLink: {
          template: '<div class="router-link-exact-active"><slot /></div>'
        }
      }
    }
  })
  expect(wrapper.find('.router-link-exact-active').exists()).toBe(true)
})

it('should react to style changes when navigating', async () => {
  const wrapper = mount(Sidebar)
  // Simular navegação
  await wrapper.setData({ isVisible: true })
  expect(wrapper.vm.isVisible).toBe(true)
})

it('should render without errors with default props', () => {
  const wrapper = mount(Sidebar)
  expect(wrapper.html()).toMatchSnapshot()
})
