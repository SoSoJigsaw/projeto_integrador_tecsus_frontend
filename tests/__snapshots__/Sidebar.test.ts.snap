// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar.vue > should render without errors with default props 1`] = `
"<!-- Menu lateral -->
<div class="sidebar">
  <div class="logo"><img src="/src/assets/logo/tecsus.png" class="logo_img" style=""></div>
  <nav class="links">
    <router-link-stub to="/"></router-link-stub>
    <router-link-stub to="/reports"></router-link-stub>
    <router-link-stub to="/alerts"></router-link-stub>
    <router-link-stub to="/importar-dados"></router-link-stub>
  </nav>
</div>"
`;
